scripts/trim_indel_ends.py

```diff
@@ -0,0 +1,402 @@
+
+def extract_edits(cigar_str) -> list:
+    """Given a cigar string, return separated edits with counts
+
+    Args:
+        cigar_str (str):
+
+    Returns:
+        list: a list of tuples of (count, edit_type)
+    """
+    all_edits = [(int(s[:-1]), s[-1]) for s in re.findall(r'\d+[=,I,D,X]', cigar_str)]
+    return all_edits
+
+def ed_from_cigar(cigar_str) -> int:
+    """Compute edit distance from cigar strings
+
+    Args:
+        cigar_str (str): cigar string
+
+    Returns:
+        int: edit distance that is equal to total length of the string minus the nu
+    """
+    all_edits = extract_edits(cigar_str)
+    total_length = sum([e[0] for e in all_edits])
+    match_count = sum([e[0] for e in all_edits if e[1] == "="])
+    return total_length - match_count
+
+def get_len_type(indel_list:list) -> tuple[str, str]:
+    """Given a list of indels, return the total length and type
+
+    Args:
+        indel_list (list): list of indels
+
+    Returns:
+        tuple[str, str]: type, length
+    """
+
+    indel_type = ""
+    indel_length = 0
+    if len(indel_list) != 0:
+        for edit in indel_list:
+            if indel_type != "":
+                assert indel_type == edit[-1]
+            else:
+                indel_type = edit[-1]
+            indel_length += edit[0]
+    return indel_type, indel_length
+
+def get_prefix_and_suffix_indels(cigar_str:str, mode="lenient") -> tuple:
+    """Get indel edits from the prefix and suffix of the cigar string
+
+    Args:
+        cigar_str (str): cigar string
+
+    Returns:
+        tuple: (list, list): prefix list of indels and suffix list of indels
+    """
+    all_edits = extract_edits(cigar_str)
+    prefix_indels = []
+    suffix_indels = []
+    i = 0
+    for i in range(len(all_edits)):
+        if all_edits[i][-1] != "I" and all_edits[i][-1] != "D":
+            break
+        prefix_indels.append(all_edits[i])
+    j = len(all_edits)-1
+    for j in range(len(all_edits)-1, -1, -1):
+        if all_edits[j][-1] != "I" and all_edits[j][-1] != "D":
+            break
+        suffix_indels.append(all_edits[j])
+    for idx in range(i,j+1):
+        if all_edits[idx][-1] != "=" and mode == "strict":
+            return ("", 0), ("",0)
+
+    # return prefix_indels, suffix_indels[::-1]
+    return get_len_type(prefix_indels), get_len_type(suffix_indels[::-1])
+
+def trim_seq(prefix_type, prefix_len, suffix_type, suffix_len, seq, num, threshold=10) -> str:
+    """Trim input sequence according to suffix and prefix number of indels
+
+    Args:
+        prefix_type (str): type of edits at the beginning of the cigar, ("I" or "D")
+        prefix_len (int): length of prefix edits
+        suffix_type (str):type of edits at the end of the cigar, ("I" or "D")
+        suffix_len (int): length of suffix edits
+        seq (str): input sequence to be trimmed
+        num (int): ranking of the sequence. If a sequence is the first in alignment, deletion means that removing. 
```

Comment:
I didn't follow this sentence

Reply:
The edits in the cigar string refers to changes that occur on the first string to make the two strings equal. 
For example, given two strings "AATCGC" and "ATCGCC", the cigar string would be "1D5=1I", which means deleting the first character in string 1 and adding "C" in string 1. When we trim the two strings, we are in fact trimming the first character in string 1 according to "1D" and the last character in string 2 according to "1I".

The parameter  "num" tells the function whether the string is the first of the second string in this particular alignment so the function knows whether to trim according to "D" or "I".

Reply:
Thanks, this makes sense to me now